nitro-src/claim.ts

```diff
@@ -1,163 +1,219 @@
-import { AllocationType, Exit, SingleAssetExit } from "../src/types";
-
+import {
+  Allocation,
+  AllocationType,
+  Exit,
+  SingleAssetExit,
+} from "../src/types";
+import { constants } from "ethers";
+import _ from "lodash";
+
+/**
+ * Note about the inputs.
+ * exitRequest is a 2 dimensional array.
+ * - The first index enumerates assets.
+ * - The second index enumerates destinations.
+ * This means that the target channel must have the following characteristics.
+ * - Every asset must list every destination.
+ * - The destination order must be consistent across assets.
+ */
+/**
+ * @dev Computes the new outcome that should be stored against a target channel after a claim is made on its guarantor.
+ * @param guarantees the outcome containing at least one guarantee(s) which will be claimed for each asset.
+ * @param holdings initial quantity of each asset held on chain for the guarantor channel. Order matches that of initialGuaranteeOutcome.
+ * @param targetChannelIndex the index of the guarantee in the list of guarantees for the given asset -- equivalent to declaring a target channel
+ * @param targetOutcome initial outcome stored on chain for the target channel.
+ * @param exitRequest list of indices expressing which destinations in the allocation should be paid out for each asset.
+ */
 export function claim(
-  initialGuaranteeOutcome: Exit,
-  initialHoldings: BigNumberish[],
+  guarantees: Exit,
+  holdings: BigNumber[],
   targetChannelIndex: number,
-  initialTargetOutcome: Exit,
+  targetOutcome: Exit,
   exitRequest: number[][]
 ) {
-  if (initialTargetOutcome.length !== initialHoldings.length) throw Error;
-  const updatedTargetOutcome: Exit = [];
-  // We want to create a clone of the original values to prevent mutation
-  const updatedHoldings = initialHoldings.map((h) => BigNumber.from(h));
-  const updatedGuaranteeOutcome: Exit = JSON.parse(
-    JSON.stringify(initialGuaranteeOutcome)
-  );
-  const exit: Exit = [];
+  if (targetOutcome.length !== holdings.length) {
+    throw Error("targetOutcome length must match funds length");
+  }
 
-  // Iterate through every asset
-  for (
-    let assetIndex = 0;
-    assetIndex < initialGuaranteeOutcome.length;
-    assetIndex++
-  ) {
-    const guarantees = initialGuaranteeOutcome[assetIndex].allocations;
-    const targetAllocations = initialTargetOutcome[assetIndex].allocations;
+  // Deep-copy as we will mutating these objects.
+  const afterClaimGuarantee: Exit = _.cloneDeep(guarantees);
+  const afterClaimHoldings = _.cloneDeep(holdings);
+
+  const afterClaimTargetOutcome: Exit = [];
+  const afterClaimExits: Exit = [];
 
-    const updatedAllocations = targetAllocations.map((a) =>
-      JSON.parse(JSON.stringify(a))
-    ); // copy the allocations for mutation
+  // Iterate through every asset
+  for (let assetIndex = 0; assetIndex < guarantees.length; assetIndex++) {
+    const guaranteesForOneAsset = guarantees[assetIndex].allocations;
```

Comment:
```suggestion
    const allocationsForOneAsset = guarantees[assetIndex].allocations;
```
Is it better to be less specific here? We expect the allocation at `targetChannelIndex` to be a guarantee, but not all of the allocations.

The same question exists for the function signature of `claim`.

Reply:
That's a good point. I tried to not alter the functionality of `claim` in this refactor (unless I thought there was a bug). We likely do want to remove the requirement that every allocation for an asset is a guarantee.